lib/src/embedded/isolate_dispatcher.dart

```diff
@@ -105,25 +116,41 @@ class IsolateDispatcher {
     } else {
       var future = ReusableIsolate.spawn(_isolateMain);
       isolate = await future;
+      isolate.receivePort.listen((message) {
```

Comment:
Why is it necessary to pull this logic out here, rather than keeping it in `ReusableIsolate`?

Reply:
From previous implementation:

```
        // Parse out the compilation ID and surface the [ProtocolError] as an
        // error. This allows the [IsolateDispatcher] to notice that an error
        // has occurred and close out the underlying channel.

        // Protocol errors have already been through [_handleError] in the child
        // isolate, so we just send them as-is and close out the underlying
        // channel.
```

First, the motivation to remove the intermediate `StreamController` is to eliminate two unnecessary performance overheads:
1. To eliminate the overhead of deserializing `ProtocolError` packet, passing it through `addError`, and re-serializing it back to packet. - We don't actually care about the `ProtocolError` object, all we care is that the category 2 indicate the packet contains a ProtocolError.
2. To eliminate the overhead of allocating new `StreamController` for every single compilation.

In regarding of why the `receivePort.listen` is moved to `IsolateDispatcher`, IMO logically how the message is sent is a contract between `IsolateDispatcher` and `CompilationDispatcher`, `ReusableIsolate` is should purely be a passthrough helper just to help with creating and managing isolate for repeated use.

Reply:
I think you're right that putting the decoding logic here makes the most sense. But I'm not a big fan of giving the caller responsibility for maintaining a single stream subscription across multiple borrows, and I'm not very worried about the overhead of an extra `StreamController`. Can we make `.borrow()` return a subscription that just forwards a view of the `Stream<object?>` which closes automatically once `.return()` is called?

Reply:
Not precisely a new `Stream` each time, but I made `.borrow` take a new `onData` function each time, which behaves somewhat like the previous implementation but without creating new `StreamController`.